### ISSUE TITLE
Modifying instance value affects the default one

### ISSUE THREAD

**Issue by DawidJanczak** [dry-rb/dry-struct#104]
Hello,

We spotted a weird behaviour when modifying a value of a struct instance where the new value gets picked up as the default:
```ruby
class Foo < Dry::Struct
  attribute :array, Types::Strict::Array.of(Types::Strict::Integer).default([])  
end

foo = Foo.new
p foo.array # []
foo.array << 1
p foo.array # [1]

bar = Foo.new
p bar.array # [1]
```
We worked around it by using `.new` on the instance (as we probably should have in the beginning), but I wonder if this isn't a bug regardless as I would have thought `<<` would only affect the instance.

**Comment by AMHOL** [comment]
@DawidJanczak I don't think this is a bug, if you pass a non-callable to `#default` `dry-types` will just set that value as the default. 

You can pass a block to `#default` which will be `#call`ed, and the return value set as the default value each time you use a type:

```ruby
type = Types::Strict::Array.of(Types::Strict::Integer).default { [] }

one = type.()
# => []
one << 1
# => [1]
two = type.()
# => []
```

So in your case it's just:

```ruby
class Foo < Dry::Struct
  attribute :array, Types::Strict::Array.of(Types::Strict::Integer).default { [] }
end
```

Will leave open for now to get opinions from core team.

**Comment by DawidJanczak** [comment]
Thank you both!
